src/hooks/useTransition.js

```diff
@@ -0,0 +1,58 @@
+import React from 'react'
+import { usePrevious } from './usePrevious'
+
+const UNMOUNTED = 'unmounted'
+const MOUNTING = 'mounting'
+const MOUNTED = 'mounted'
+const UNMOUNTING = 'unmounting'
+
+export function useTransition({ on = false, duration = 500 }) {
+  const [state, setState] = React.useState(UNMOUNTED)
+  const [show, setShow] = React.useState(false)
+  const [transitioning, setTransitioning] = React.useState(false)
+  const prevOn = usePrevious(on)
+
+  React.useEffect(() => {
+    function handleMount() {
+      if (transitioning) return false
+      setState(MOUNTING)
+      function tick() {
+        setState(MOUNTED)
+        setTransitioning(false)
+      }
+      let id = setTimeout(tick, duration)
+      return () => clearTimeout(id)
+    }
+
+    function handleUnmount() {
+      if (transitioning) return false
+      setTransitioning(true)
+      setState(UNMOUNTING)
+      function tick() {
+        setState(UNMOUNTED)
+        setShow(false)
+        setTransitioning(false)
+      }
+      let id = setTimeout(tick, duration)
+      return () => clearTimeout(id)
+    }
+
+    if (on && !prevOn) {
+      setTransitioning(true)
+      setShow(true)
+      requestAnimationFrame(() => {
+        requestAnimationFrame(handleMount)
```

Comment:
Can you explain the need for 2 x `requestAnimationFrame`?

Reply:
https://gist.github.com/getify/3002042#gistcomment-2351663

If you’d only call it once, then the initial transition after initial page-load won’t work. It’s bit weird, but you can try it out by removing one call, refresh the page, and open a dialog. Alternatively a `setTimeout()` call with `0` delay also works, but I thought it was uglier than the above.